beacon-chain/rpc/beaconv1/validator.go

```diff
@@ -2,27 +2,165 @@ package beaconv1
 func (bs *Server) ListValidators(ctx context.Context, req *ethpb.StateValidatorsRequest) (*ethpb.StateValidatorsResponse, error) {
-	return nil, errors.New("unimplemented")
+	state, err := bs.StateFetcher.State(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state: %v", err)
+	}
+
+	valContainers := make([]*ethpb.ValidatorContainer, len(req.Id))
+	for i := 0; i < len(req.Id); i++ {
+		valContainer, err := valContainerById(state, req.Id[i])
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get validator container: %v", err)
+		}
+		valContainers[i] = valContainer
+	}
+	return &ethpb.StateValidatorsResponse{Data: valContainers}, nil
 }
 
 // ListValidatorBalances returns a filterable list of validator balances.
 func (bs *Server) ListValidatorBalances(ctx context.Context, req *ethpb.ValidatorBalancesRequest) (*ethpb.ValidatorBalancesResponse, error) {
-	return nil, errors.New("unimplemented")
+	state, err := bs.StateFetcher.State(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state: %v", err)
+	}
+
+	valBalances := make([]*ethpb.ValidatorBalance, len(req.Id))
+	for i := 0; i < len(req.Id); i++ {
+		_, valIndex, err := validatorById(state, req.Id[i])
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get validator: %v", err)
+		}
+		bal, err := state.BalanceAtIndex(valIndex)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get balance at index %d: %v", valIndex, err)
+		}
+		valBalances[i] = &ethpb.ValidatorBalance{
+			Index:   valIndex,
+			Balance: bal,
+		}
+	}
+	return &ethpb.ValidatorBalancesResponse{Data: valBalances}, nil
 }
 
 // ListCommittees retrieves the committees for the given state at the given epoch.
+// If the requested slot and index are defined, only those committees are returned.
 func (bs *Server) ListCommittees(ctx context.Context, req *ethpb.StateCommitteesRequest) (*ethpb.StateCommitteesResponse, error) {
-	return nil, errors.New("unimplemented")
+	state, err := bs.StateFetcher.State(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state: %v", err)
+	}
+
+	epoch := helpers.SlotToEpoch(state.Slot())
+	if req.Epoch != nil {
+		epoch = *req.Epoch
+	}
+	activeCount, err := helpers.ActiveValidatorCount(state, epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get active validator count: %v", err)
+	}
+
+	startSlot, err := helpers.StartSlot(epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get epoch start slot: %v", err)
+	}
+	endSlot, err := helpers.EndSlot(epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get epoch end slot: %v", err)
+	}
+	committeesPerSlot := helpers.SlotCommitteeCount(activeCount)
+	committees := make([]*ethpb.Committee, 0)
+	for slot := startSlot; slot <= endSlot; slot++ {
```

Comment:
The logic of this loop is incorrect. We must return all indexes/slots when the corresponding parameter is not provided. So in pseudocode this would be:

```
if index_provided {
  if slot_provided {
    get_committee(index, slot)
  } else {
    forall s in slots get_committee(index, s)
  }
} else {
  forall i in committee_indices 
    if slot_provided {
      get_committee(i, slot)
    } else {
      forall s in slots get_committee(i, s)
    }
}

Again it would be nice to have some optimizations, although the number of loops is a few magnitudes lower than for validators and is unlikely to increase substantially, so it doesn't matter that much.

Reply:
How is the logic incorrect? Doesn't it return everything for the head epoch when there is nothing provided in the request?

Reply:
You are right, I did not correctly analyze your code.